source/SpinalHDL/Libraries/Pipeline/introduction.rst

```diff
@@ -0,0 +1,794 @@
+      wire                nodes_0_valid;
+      reg                 nodes_0_ready;
+      reg                 nodes_1_valid;
+      reg                 nodes_1_ready;
+      reg                 nodes_2_valid;
+      reg                 nodes_2_ready;
+      reg                 nodes_3_valid;
+      wire                nodes_3_ready;
+      wire                when_StageConnector_l56;
+      wire                when_StageConnector_l56_1;
+      wire                when_StageConnector_l56_2;
+
+      assign _zz_nodes_0_adder_SUM = (nodes_0_inserter_RGB_r + nodes_0_inserter_RGB_g);
+      assign nodes_0_valid = io_up_valid;
+      assign io_up_ready = nodes_0_ready;
+      assign nodes_0_inserter_RGB_r = io_up_payload_r;
+      assign nodes_0_inserter_RGB_g = io_up_payload_g;
+      assign nodes_0_inserter_RGB_b = io_up_payload_b;
+      assign nodes_0_adder_SUM = (_zz_nodes_0_adder_SUM + nodes_0_inserter_RGB_b);
+      assign nodes_1_inverter_INV = (~ nodes_1_adder_SUM);
+      assign nodes_2_multiplier_MUL = (nodes_2_inverter_INV * 8'hee);
+      assign io_down_valid = nodes_3_valid;
+      assign nodes_3_ready = io_down_ready;
+      assign io_down_payload = nodes_3_multiplier_MUL;
+      always @(*) begin
+        nodes_0_ready = nodes_1_ready;
+        if(when_StageConnector_l56) begin
+          nodes_0_ready = 1'b1;
+        end
+      end
+
+      assign when_StageConnector_l56 = (! nodes_1_valid);
+      always @(*) begin
+        nodes_1_ready = nodes_2_ready;
+        if(when_StageConnector_l56_1) begin
+          nodes_1_ready = 1'b1;
+        end
+      end
+
+      assign when_StageConnector_l56_1 = (! nodes_2_valid);
+      always @(*) begin
+        nodes_2_ready = nodes_3_ready;
+        if(when_StageConnector_l56_2) begin
+          nodes_2_ready = 1'b1;
+        end
+      end
+
+      assign when_StageConnector_l56_2 = (! nodes_3_valid);
+      always @(posedge clk or posedge reset) begin
+        if(reset) begin
+          nodes_1_valid <= 1'b0;
+          nodes_2_valid <= 1'b0;
+          nodes_3_valid <= 1'b0;
+        end else begin
+          if(nodes_0_ready) begin
+            nodes_1_valid <= nodes_0_valid;
+          end
+          if(nodes_1_ready) begin
+            nodes_2_valid <= nodes_1_valid;
+          end
+          if(nodes_2_ready) begin
+            nodes_3_valid <= nodes_2_valid;
+          end
+        end
+      end
+
+      always @(posedge clk) begin
+        if(nodes_0_ready) begin
+          nodes_1_adder_SUM <= nodes_0_adder_SUM;
+        end
+        if(nodes_1_ready) begin
+          nodes_2_inverter_INV <= nodes_1_inverter_INV;
+        end
+        if(nodes_2_ready) begin
+          nodes_3_multiplier_MUL <= nodes_2_multiplier_MUL;
+        end
+      end
+
+
+    endmodule
+
+
+Also, you can easily tweek how many stages and where you want the processing to be done, for instance you may want to move the invertion hardware in the same stage as the adder. This can be done the following way : 
```

Comment:
```suggestion
Also, you can easily tweak how many stages and where you want the processing to be done, for instance you may want to move the inversion hardware in the same stage as the adder. This can be done the following way : 
```